### PR TITLE
Rework PCE

### LINKED ISSUE #2894: Add PCE strategy based on latency
# Scope
1. Add a PCE strategy that takes ISL latency as the main weight parameter instead of ISL cost. 
1. Add the ability to specify the PCE strategy on flow and system level.

### PR DESCRIPTION
**Pull request by rozdy**
Add path computation strategy param
Rework path find algorithm to work with abstract weight function
Add path computation strategy based on latency

Closes #2894

Renamed properties:
isl.cost.when.port.down -> pce.isl.cost.when.port.down
isl.cost.when.under.maintenance -> pce.isl.cost.when.under.maintenance
pce.diversity.isl.weight -> pce.diversity.isl.cost
pce.diversity.switch.weight -> pce.diversity.switch.cost

---
This change is [Reviewable](https://reviewable.io/reviews/telstra/open-kilda/2910)

### CODE REVIEW

**Comment by rozdy** [comment]
> > Reviewed 54 of 54 files at r1.
> > _[Reviewable](https://reviewable.io/reviews/telstra/open-kilda/2910)_ status: all files reviewed, 2 unresolved discussions (waiting on @dpoltavets, @niksv, @rozdy, @sergii-iakovenko, @surabujin, and @timofei-durakov)
> 
> _[services/src/kilda-pce/src/main/java/org/openkilda/pce/impl/InMemoryPathComputer.java, line 128 at r1](https://reviewable.io/reviews/telstra/open-kilda/2910#-LteiRTD58czjeVFY964:-LteiRTD58czjeVFY965:b-hjy0pw) ([raw file](https://github.com/telstra/open-kilda/blob/79e4f0853d4f0047496063599548464d03dbf51b/services/src/kilda-pce/src/main/java/org/openkilda/pce/impl/InMemoryPathComputer.java#L128)):_
> 
> > ```java
> >         switch (strategy) {
> >             case COST:
> >                 return edge -> {
> > ```
> 
> move to method
> 
> _[services/src/messaging/src/main/java/org/openkilda/messaging/payload/flow/PathComputationStrategy.java, line 21 at r1](https://reviewable.io/reviews/telstra/open-kilda/2910#-Ltej5-9AO2dewmzJKGq:-Ltej5-A2YlreDXw9-eP:bom83qu) ([raw file](https://github.com/telstra/open-kilda/blob/79e4f0853d4f0047496063599548464d03dbf51b/services/src/messaging/src/main/java/org/openkilda/messaging/payload/flow/PathComputationStrategy.java#L21)):_
> 
> > ```java
> >  * Class represents path computation strategy.
> >  */
> > public enum PathComputationStrategy {
> > ```
> 
> don't duplicate

Fixed

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowCreateService.java [hunk 1]
```diff
@@ -73,6 +73,9 @@ public void handleRequest(String key, CommandContext commandContext, FlowRequest
         if (requestedFlow.getFlowEncapsulationType() == null) {
             requestedFlow.setFlowEncapsulationType(kildaConfigurationRepository.get().getFlowEncapsulationType());
         }
+        if (requestedFlow.getPathComputationStrategy() == null) {
+            requestedFlow.setPathComputationStrategy(kildaConfigurationRepository.get().getPathComputationStrategy());
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowCreateService.java]: Please, don't forget to make the same in FlowUpdateService.

File: services/src/kilda-core/kilda-model/src/main/java/org/openkilda/model/KildaConfiguration.java [hunk 2]
```diff
@@ -53,13 +54,19 @@
     @Property(name = "use_multi_table")
     private Boolean useMultiTable;
 
+    @Property(name = "path_computation_strategy")
```
[the remark this answers is not in the document]
Comment (rozdy) [reply services/src/kilda-core/kilda-model/src/main/java/org/openkilda/model/KildaConfiguration.java]: I guess no. All values in Kilda Configuration are default values for the system.

File: confd/templates/wfm/topology.properties.tmpl [hunk 3]
```diff
@@ -70,10 +68,15 @@ floodlight.regions = {{ getv "/kilda_floodlight_regions" }}
...
 pce.network.strategy = {{ getv "/kilda_pce_network_strategy" }}
-pce.diversity.isl.weight = {{ getv "/kilda_diversity_isl_weight" }}
-pce.diversity.switch.weight = {{ getv "/kilda_diversity_switch_weight" }}
+pce.diversity.isl.cost = {{ getv "/kilda_diversity_isl_cost" }}
+pce.diversity.switch.cost = {{ getv "/kilda_diversity_switch_cost" }}
+pce.isl.cost.when.port.down = {{ getv "/kilda_isl_cost_when_port_down" }}
+pce.isl.cost.when.under.maintenance = {{ getv "/kilda_isl_cost_when_under_maintenance" }}
+pce.isl.latency.when.port.down = {{ getv "/kilda_isl_latency_when_port_down" }}
+pce.isl.latency.when.under.maintenance = {{ getv "/kilda_isl_latency_when_under_maintenance" }}
+pce.diversity.isl.latency = {{ getv "/kilda_diversity_isl_latency" }}
```
[the remark this answers is not in the document]
Comment (rozdy) [reply confd/templates/wfm/topology.properties.tmpl]: Added.

File: services/src/kilda-core/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/converters/PathComputationStrategyConverter.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        }
+        return value.name();
+    }
+
+    @Override
+    public PathComputationStrategy toEntityAttribute(String value) {
+        if (value == null || value.trim().isEmpty()) {
+            return null;
+        }
+        return PathComputationStrategy.valueOf(value);
```
[the remark this answers is not in the document]
Comment (rozdy) [reply services/src/kilda-core/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/converters/PathComputationStrategyConverter.java]: Done.

File: services/src/kilda-pce/src/main/java/org/openkilda/pce/PathComputerConfig.java [hunk 5]
```diff
@@ -32,19 +32,43 @@
...
-    @Key("strategy")
-    @Default("COST")
-    String getStrategy();
+    int getDiversitySwitchCost();
 
     @Key("network.strategy")
     @Default("COST")
     String getNetworkStrategy();
+
+    @Key("isl.cost.when.port.down")
```
[the remark this answers is not in the document]
Comment (rozdy) [reply services/src/kilda-pce/src/main/java/org/openkilda/pce/PathComputerConfig.java]: Renamed.

File: services/src/kilda-pce/src/main/java/org/openkilda/pce/PathComputerConfig.java [hunk 6]
```diff
@@ -32,19 +32,43 @@
...
+    @Default("1000000000")
+    long getDiversityIslLatency();
+
+    @Key("diversity.switch.latency")
+    @Default("300000000")
+    long getDiversitySwitchLatency();
+
+    @Key("isl.latency.when.port.down")
+    @Default("10000000000")
+    long getUnstableLatencyRaise();
```
[the remark this answers is not in the document]
Comment (rozdy) [reply services/src/kilda-pce/src/main/java/org/openkilda/pce/PathComputerConfig.java]: Renamed.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowCreateService.java (hunk 1 above)
Comment (rozdy) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowCreateService.java]: Rebased and added.

File: services/src/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java [hunk 7]
```diff
@@ -142,24 +149,24 @@ public void shouldReduceWithDiversity() {
...
+                weightFunction.apply(network.getSwitch(SRC_SWITCH).getOutgoingLinks().iterator().next()));
+        assertEquals(expectedWeight,
+                weightFunction.apply(network.getSwitch(DST_SWITCH).getOutgoingLinks().iterator().next()));
 
-        assertEquals(cost,
-                network.getSwitch(DST_SWITCH).getOutgoingLinks().iterator().next().getFullWeight(weightFunction));
-        assertEquals(cost,
-                network.getSwitch(DST_SWITCH).getIncomingLinks().iterator().next().getFullWeight(weightFunction));
+        assertEquals(expectedWeight,
+                weightFunction.apply(network.getSwitch(DST_SWITCH).getOutgoingLinks().iterator().next()));
```
[the remark this answers is not in the document]
Comment (rozdy) [reply services/src/kilda-pce/src/test/java/org/openkilda/pce/impl/AvailableNetworkTest.java]: Fixed this test. We already have test case for asymmetric forward and reverse paths: `testForwardAndBackwardPathsEqualityEvenWhenReverseHasCheaperPath()`.

**Pull request merged**